apps/voting/app/src/vote-hooks.js

```diff
@@ -0,0 +1,79 @@
+import { useState, useEffect } from 'react'
+
+// when `params` are updated, call `fn` and pass the result
+function usePromiseResult(fn, params, memoParams, defaultValue) {
+  const [result, setResult] = useState(defaultValue)
+  useEffect(() => {
+    let cancelled = false
+    fn(...params).then(value => {
+      if (!cancelled) {
+        setResult(value)
+      }
+    })
+    return () => {
```

Comment:
I like the idea of cancelling promises but this mechanism is not being used is it? Do we need it?

Reply:
We are not cancelling the promise here (even though it could be a good idea to implement it this way), but it is important to prevent the state to be updated if it gets called with another `memoParams`. That way we don’t get into concurrency bugs if e.g. it gets called twice but the response of the first call comes after the second one.

We need it because in this case, because `.call()` is asynchronous.